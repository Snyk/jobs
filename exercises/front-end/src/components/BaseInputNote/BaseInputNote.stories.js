import '~/assets/scss/styles.scss';
import BaseInputNote from './BaseInputNote';

export default {
  title: 'Library/Base Components/InputNote',
  component: BaseInputNote,
};

export const Default = () => ({
  components: { BaseInputNote },
  props: {
    variant: {
      default: 'info',
    },
  },
  template: `
    <BaseInputNote :variant="variant">
      Base Input Note ({{variant}}).
    </BaseInputNote>
  `,
});

export const Info = () => ({
  components: { BaseInputNote },
  template: `<BaseInputNote variant="info">A info input note for informational purposes.</BaseInputNote>`,
});

export const Warning = () => ({
  components: { BaseInputNote },
  template: `<BaseInputNote variant="warning">A warning input note for when there's a non-critical problem.</BaseInputNote>`,
});

export const Success = () => ({
  components: { BaseInputNote },
  template: `<BaseInputNote variant="success">A success input note for when something awesome has happened.</BaseInputNote>`,
});

export const Danger = () => ({
  components: { BaseInputNote },
  template: `<BaseInputNote variant="danger">A danger input note for when something unpleasant has happened.</BaseInputNote>`,
});
