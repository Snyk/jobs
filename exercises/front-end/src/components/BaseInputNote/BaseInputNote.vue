<template>
  <div class="vue--input-note" :class="computedClasses">
    <slot />
  </div>
</template>

<script>
import { isInList } from '~/lib/prop-validators';

/**
 * A note to appear under an input
 */
export default {
  name: 'BaseInputNote',

  status: 'under-review',

  props: {
    /**
     * Variant of BaseInputNote. Must be one of `default`, `info`, `warning`,
     * `error`, `danger`, `success`. Defaults to `default`
     */
    variant: {
      type: String,
      default: 'info',
      validator: isInList([
        'default',
        'info',
        'warning',
        'error',
        'danger',
        'success',
      ]),
    },
  },

  computed: {
    computedClasses() {
      return {
        [`vue--input-note--${this.variant}`]: true,
      };
    },
  },
};
</script>

<style lang="scss" scoped>
@import '~/assets/scss/utils';

.vue--input-note {
  @include typeset(caption);

  &--default,
  &--info {
    color: color(base, dimmed);
  }

  &--warning {
    color: color(alert, warning-text);
  }

  &--error,
  &--danger {
    color: color(alert, danger-text);
  }

  &--success {
    color: color(alert, success-text);
  }
}
</style>
