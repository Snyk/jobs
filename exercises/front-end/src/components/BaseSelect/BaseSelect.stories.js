import '~/assets/scss/styles.scss';
import BaseSelect from './BaseSelect';

export default {
  title: 'Library/Base Components/Select',
  component: BaseSelect,
};

export const Default = () => ({
  components: { BaseSelect },
  props: {
    label: {
      default: 'Default label',
    },
    options: {
      default: [
        { title: 'Option 1', value: '1' },
        { title: 'Option 2', value: '2' },
        { title: 'Option 3', value: '3' },
      ],
    },
    size: {
      default: null,
    },
    error: {
      default: false,
    },
    fillWidth: {
      default: false,
    },
    inline: {
      default: false,
    },
    note: {
      default: '',
    },
  },
  data() {
    return {
      value: '1',
    };
  },
  template: `
    <BaseSelect
      :label="label"
      :options="options"
      :size="size"
      :error="error"
      :fill-width="fillWidth"
      :inline="inline"
      v-model="value"
      :note="note"
    />
  `,
});

export const FillWidth = () => ({
  components: { BaseSelect },
  props: {
    selectOptions: {
      default: [
        {
          value: '1',
          title: 'Option 1',
        },
        {
          value: '2',
          title: 'Option 2',
        },
        {
          value: '3',
          title: 'Option 3',
        },
      ],
    },
  },
  template: `
  <BaseSelect
    label="Fill width select"
    :options="selectOptions"
    fill-width
  />
  `,
});
FillWidth.storyName = '…as full width';

export const States = () => ({
  components: { BaseSelect },
  props: {
    selectOptions: {
      default: [
        {
          value: '1',
          title: 'Option 1',
        },
        {
          value: '2',
          title: 'Option 2',
        },
        {
          value: '3',
          title: 'Option 3',
        },
      ],
    },
  },
  template: `
    <div>
      <BaseSelect
        label=":hover"
        state="hover"
        :options="selectOptions"
      />
      <BaseSelect
        label=":docus"
        state="focus"
        :options="selectOptions"
      />
      <BaseSelect
        label="[disabled]"
        :options="selectOptions"
        disabled
      />
      <BaseSelect
        label="Error"
        :options="selectOptions"
        error
      />
    </div>
  `,
});
States.storyName = '…with states';

export const Sizes = () => ({
  components: { BaseSelect },
  props: {
    selectOptions: {
      default: [
        {
          value: '1',
          title: 'Option 1',
        },
        {
          value: '2',
          title: 'Option 2',
        },
        {
          value: '3',
          title: 'Option 3',
        },
      ],
    },
  },
  template: `
    <div>
      <BaseSelect
        label="Default"
        :options="selectOptions"
      />
      <BaseSelect
        label="Small"
        :options="selectOptions"
        size="small"
      />
    </div>
  `,
});
Sizes.storyName = '…with sizes';

export const Inline = () => ({
  components: { BaseSelect },
  props: {
    selectOptions: {
      default: [
        {
          value: '1',
          title: 'Option 1',
        },
        {
          value: '2',
          title: 'Option 2',
        },
        {
          value: '3',
          title: 'Option 3',
        },
      ],
    },
  },
  template: `
    <BaseSelect
      label="Inline"
      :options="selectOptions"
      inline
    />
  `,
});
Inline.storyName = '…as inline';

export const PreSelected = () => ({
  components: { BaseSelect },
  props: {
    selectOptions: {
      default: [
        {
          value: '1',
          title: 'Option 1',
        },
        {
          value: '2',
          title: 'Option 2',
        },
        {
          value: '3',
          title: 'Option 3',
        },
      ],
    },
  },
  data() {
    return {
      value: '2',
    };
  },
  template: `
    <BaseSelect
      :options="selectOptions"
      v-model="value"
    />
  `,
});
PreSelected.storyName = '…as pre selected';

export const Notes = () => ({
  components: { BaseSelect },
  props: {
    selectOptions: {
      default: [
        {
          value: '1',
          title: 'Option 1',
        },
        {
          value: '2',
          title: 'Option 2',
        },
        {
          value: '3',
          title: 'Option 3',
        },
      ],
    },
  },
  template: `
  <div>
    <BaseSelect
      label="With note"
      :options="selectOptions"
      note="Passwords must contain at least one capital letter, a number, and a special character."
    />
    <BaseSelect
      label="Error with note"
      :options="selectOptions"
      error
      note="This is a required field"
    />
  </div>
  `,
});
Notes.storyName = '…with note';

export const MinimalVariant = () => ({
  components: { BaseSelect },
  props: {
    options: {
      default: [
        { title: 'Option 1', value: '1' },
        { title: 'Option 2', value: '2' },
        { title: 'Option 3', value: '3' },
      ],
    },
  },
  data() {
    return {
      value: '1',
    };
  },
  template: `
    <BaseSelect
      variant="minimal"
      :options="options"
      v-model="value"
    />
  `,
});

MinimalVariant.storyName = '…as minimal variant';
export const InvertedMinimalVariant = () => ({
  components: { BaseSelect },
  props: {
    options: {
      default: [
        { title: 'Option 1', value: '1' },
        { title: 'Option 2', value: '2' },
        { title: 'Option 3', value: '3' },
      ],
    },
  },
  data() {
    return {
      value: '1',
    };
  },
  template: `
    <div style="background-color: hsl(244, 8%, 24%);">
      <BaseSelect
        variant="minimal-inverted"
        :options="options"
        v-model="value"
      />
    </div>
  `,
});
InvertedMinimalVariant.storyName = '…as inverted minimal variant';
