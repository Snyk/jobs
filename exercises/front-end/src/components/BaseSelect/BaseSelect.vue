<template>
  <LayoutFormElement
    class="vue--select"
    :class="computedWrapperClasses"
    :inline="inline"
    :fill-width="fillWidth"
  >
    <template v-if="label" #label>
      <label class="vue--select__label" :for="$attrs.id" v-text="label" />
    </template>
    <div class="vue--select__container">
      <select
        v-bind="$attrs"
        :class="computedInputClasses"
        class="vue--select__field"
        v-on="listeners"
      >
        <option
          v-for="option in options"
          :key="option.value"
          class="vue--select__option"
          :value="option.value"
          :selected="option.value === selected"
          :disabled="option.disabled"
          v-text="option.title"
        />
      </select>
      <ChevronDown class="vue--select__arrow" />
      <div
        v-if="showSelectedValue"
        class="vue--select__selected-value"
        v-text="selectedValueTitle"
      />
    </div>
    <BaseInputNote v-if="note" class="vue--select__note" :variant="noteVariant">
      {{ note }}
    </BaseInputNote>
  </LayoutFormElement>
</template>

<script>
import ChevronDown from 'icons/ChevronDown';
import LayoutFormElement from '~/components/LayoutFormElement/LayoutFormElement';
import BaseInputNote from '~/components/BaseInputNote/BaseInputNote';
import { isInList } from '~/lib/prop-validators';

export default {
  name: 'BaseSelect',

  status: 'under-review',

  components: {
    ChevronDown,
    LayoutFormElement,
    BaseInputNote,
  },

  inheritAttrs: false,

  model: {
    event: 'change',
  },

  props: {
    variant: {
      type: String,
      default: 'default',
      validator: isInList(['default', 'minimal', 'minimal-inverted']),
    },
    /**
     * Array of select options, with the following structure:
     * {
     *  {String}  [value]
     *  {String}  [title]
     *  {Boolean} [disabled] // optional
     * }
     */
    options: {
      type: Array,
      required: true,
    },
    /**
     * The value for the selected option.
     */
    value: {
      type: [String, Number],
      default: null,
    },
    /**
     * The label of the form select field.
     */
    label: {
      type: String,
      default: null,
    },
    /**
     * Whether to show an error for this field.
     * `true, false`
     */
    error: {
      type: Boolean,
      default: false,
    },
    /**
     * Manually trigger various states of the select.
     * `hover, active, focus`
     */
    state: {
      type: String,
      default: null,
      validator: isInList(['hover', 'active', 'focus']),
    },
    /**
     * BaseSelect should fill the available width.
     */
    fillWidth: {
      type: Boolean,
      default: false,
    },
    /**
     * Whether the form input field is displayed inline.
     * `true, false`
     */
    inline: {
      type: Boolean,
      default: false,
    },
    /**
     * Sizing for this input.
     * `small`
     */
    size: {
      type: String,
      default: null,
      validator: isInList(['small']),
    },
    /**
     * A note to be displayed below this input.
     */
    note: {
      type: String,
      default: null,
    },
  },

  data: function () {
    return {
      selected: this.value,
    };
  },

  computed: {
    listeners() {
      return {
        ...this.$listeners,
        // TODO: if we change 'change' to 'input' v-model
        // will work out of the box. Requires a refactor of
        // anything listening for the 'change' even however.
        change: (e) => this.$emit('change', e.target.value),
        focus: (e) => this.$emit('focus', e.target.value),
        keyup: (e) => {
          if (e.keyCode === 13) {
            this.$emit('enter');
          } else {
            this.$emit('keyup', e);
          }
        },
        keydown: (e) => this.$emit('keydown', e),
        blur: (e) => this.$emit('blur', e.target.value),
      };
    },
    computedWrapperClasses() {
      return {
        [`vue--select--${this.variant}`]: !!this.variant,
        'vue--select--fill-width': this.fillWidth,
        [`vue--select--${this.size}`]: this.size !== null,
      };
    },
    computedInputClasses() {
      const optionalClasses = {};
      if (this.state) {
        optionalClasses[this.state] = true;
      }

      return {
        ...optionalClasses,
        'vue--select__field--error': this.error,
      };
    },
    noteVariant() {
      return this.error ? 'error' : 'default';
    },
    showSelectedValue() {
      return ['minimal', 'minimal-inverted'].includes(this.variant);
    },
    selectedValueTitle() {
      const option = this.options.find(
        (option) => option.value === this.selected,
      );
      return option.title;
    },
  },

  watch: {
    value() {
      this.selected = this.value;
    },
  },

  created() {
    if (this.value === null && this.options.length > 0) {
      this.selected = this.options[0].value;
    }
  },
};
</script>

<style lang="scss" scoped>
@import '~/assets/scss/utils';

.vue--select {
  $self: &;

  &__arrow {
    background-color: color(input, bg);
    padding: 0 space(xs);
    pointer-events: none;
    position: absolute;
    right: 1px;
    z-index: 2;

    ::v-deep svg {
      color: color(input, text);
    }
  }

  &--default {
    #{$self} {
      &__container {
        align-items: center;
        display: inline-flex;
        position: relative;
      }

      &__field {
        @include fieldStyles();
        @include hoverState();

        appearance: none;
        height: auto; // Override legacy styles
        padding-right: space(xl);

        &[disabled],
        &:disabled {
          @include disabledState();

          + #{$self}__arrow {
            background-color: color(disabled, bg);

            ::v-deep svg {
              color: color(input, border);
            }
          }
        }

        &:focus:not(:disabled),
        &.focus:not(:disabled),
        &:hover:not(:disabled),
        &.hover:not(:disabled) {
          + #{$self}__arrow {
            ::v-deep svg {
              color: color(action);
            }
          }
        }

        &--error {
          @include errorState();

          + #{$self}__arrow {
            ::v-deep svg {
              color: color(action, destroy);
            }
          }
        }
      }

      // Hack: This makes no visible difference to rendering, but fixes an issue in firefox which caused
      // the select to resize by 8px when any options except the longest available was selected
      &__option {
        padding-left: 3px;
        padding-right: 5px;
      }

      &__note {
        margin-top: space(xs);
      }
    }
  }

  &--fill-width {
    #{$self}__container {
      display: flex;
    }
  }

  &--small {
    #{$self}__field {
      line-height: rem(22px);
    }

    #{$self}__label {
      line-height: rem(22px);
    }
  }

  &--minimal,
  &--minimal-inverted {
    #{$self}__container {
      align-items: center;
      display: inline-flex;
      position: relative;
    }

    #{$self}__field {
      opacity: 0;
      height: 100%;
      left: 0;
      position: absolute;
      top: 0;
      width: 100%;
    }

    #{$self}__selected-value {
      padding-right: rem(24px);
    }

    #{$self}__arrow {
      line-height: rem(24px);
      padding: 0;
      position: absolute;
      right: 0;

      ::v-deep svg {
        height: rem(24px);
        width: rem(24px);
      }
    }
  }

  &--minimal-inverted {
    #{$self}__selected-value {
      color: color(ui, white);
    }

    #{$self}__arrow {
      background-color: transparent;
      color: color(ui, white);
    }
  }
}
</style>
