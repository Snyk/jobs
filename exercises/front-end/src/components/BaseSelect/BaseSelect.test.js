import { shallowMount, mount } from '@vue/test-utils';
import BaseSelect from '~/components/BaseSelect/BaseSelect';
import BaseInputNote from '~/components/BaseInputNote/BaseInputNote';

describe('BaseSelect', () => {
  const componentStubs = { ChevronDown: true };
  const props = {
    options: [
      { value: '1', title: 'Option 1' },
      { value: '2', title: 'Option 2' },
      { value: '3', title: 'Option 3' },
    ],
    id: 'id-drop-mic',
    name: 'name-drop-mic',
  };

  describe('Snapshots', () => {
    it('renders as expected', () => {
      const wrapper = shallowMount(BaseSelect, {
        propsData: props,
        stubs: componentStubs,
      });
      expect(wrapper).toMatchSnapshot();
    });

    it('renders as minimal variant', () => {
      const wrapper = shallowMount(BaseSelect, {
        propsData: {
          ...props,
          variant: 'minimal',
        },
        stubs: componentStubs,
      });
      expect(wrapper).toMatchSnapshot();
    });

    it('renders as minimal inverted variant', () => {
      const wrapper = shallowMount(BaseSelect, {
        propsData: {
          ...props,
          variant: 'minimal-inverted',
        },
        stubs: componentStubs,
      });
      expect(wrapper).toMatchSnapshot();
    });

    it('renders valid html', () => {
      const wrapper = shallowMount(BaseSelect, {
        propsData: props,
        stubs: componentStubs,
      });

      expect(wrapper.html()).toHTMLValidate();
    });
  });

  describe('Defaults', () => {
    const wrapper = mount(BaseSelect, {
      propsData: props,
    });

    it('defaults to the first item being selected', () => {
      expect(wrapper.vm.selected).toEqual(props.options[0].value);
    });
  });

  describe('Computed Properties', () => {
    it('computes wrapper classes', () => {
      const wrapper = shallowMount(BaseSelect, {
        propsData: {
          ...props,
          disabled: true,
          fillWidth: true,
          size: 'small',
        },
        stubs: componentStubs,
      });

      expect(wrapper.vm.computedWrapperClasses).toEqual({
        'vue--select--default': true,
        'vue--select--fill-width': true,
        'vue--select--small': true,
      });
    });

    it('computes input classes', () => {
      const wrapper = shallowMount(BaseSelect, {
        propsData: {
          ...props,
          error: true,
        },
        stubs: componentStubs,
      });

      expect(wrapper.vm.computedInputClasses).toEqual({
        'vue--select__field--error': true,
      });
    });

    it('computes input classes with optional classes', () => {
      const wrapper = shallowMount(BaseSelect, {
        propsData: {
          ...props,
          state: 'hover',
        },
        stubs: componentStubs,
      });

      expect(wrapper.vm.computedInputClasses).toEqual({
        hover: true,
        'vue--select__field--error': false,
      });
    });
  });

  describe('With note', () => {
    const wrapper = shallowMount(BaseSelect, {
      propsData: {
        ...props,
        note: 'Note',
      },
    });

    it('renders as expected', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('contains BaseInputNote', () => {
      expect(wrapper.findComponent(BaseInputNote).isVisible()).toBe(true);
      expect(wrapper.vm.noteVariant).toEqual('default');
    });

    it('displays error variant', async () => {
      await wrapper.setProps({
        error: true,
      });
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.vm.noteVariant).toEqual('error');
    });

    it('does not contain BaseInputNote when note prop not set', async () => {
      await wrapper.setProps({
        note: undefined,
      });
      expect(wrapper.findComponent(BaseInputNote).exists()).toBe(false);
    });
  });

  describe('Events', () => {
    it('emits a change event', async () => {
      const wrapper = shallowMount(BaseSelect, { propsData: props });
      const chosenValue = '3';
      const select = wrapper.find('select');
      select.element.value = chosenValue;
      await select.trigger('change');
      const events = wrapper.emitted();
      expect(events.change).toHaveLength(1);
      expect(events.change[0]).toEqual([chosenValue]);
    });
  });
});
