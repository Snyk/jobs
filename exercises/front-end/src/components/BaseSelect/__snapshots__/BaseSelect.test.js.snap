// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseSelect Snapshots renders as expected 1`] = `
<layoutformelement-stub class="vue--select vue--select--default">
  <div class="vue--select__container"><select id="id-drop-mic" name="name-drop-mic" class="vue--select__field">
      <option class="vue--select__option" value="1">Option 1</option>
      <option class="vue--select__option" value="2">Option 2</option>
      <option class="vue--select__option" value="3">Option 3</option>
    </select>
    <chevrondown-stub title="Chevron Down icon" fillcolor="currentColor" size="24" class="vue--select__arrow"></chevrondown-stub>
    <!---->
  </div>
  <!---->
</layoutformelement-stub>
`;

exports[`BaseSelect Snapshots renders as minimal inverted variant 1`] = `
<layoutformelement-stub class="vue--select vue--select--minimal-inverted">
  <div class="vue--select__container"><select id="id-drop-mic" name="name-drop-mic" class="vue--select__field">
      <option class="vue--select__option" value="1">Option 1</option>
      <option class="vue--select__option" value="2">Option 2</option>
      <option class="vue--select__option" value="3">Option 3</option>
    </select>
    <chevrondown-stub title="Chevron Down icon" fillcolor="currentColor" size="24" class="vue--select__arrow"></chevrondown-stub>
    <div class="vue--select__selected-value">Option 1</div>
  </div>
  <!---->
</layoutformelement-stub>
`;

exports[`BaseSelect Snapshots renders as minimal variant 1`] = `
<layoutformelement-stub class="vue--select vue--select--minimal">
  <div class="vue--select__container"><select id="id-drop-mic" name="name-drop-mic" class="vue--select__field">
      <option class="vue--select__option" value="1">Option 1</option>
      <option class="vue--select__option" value="2">Option 2</option>
      <option class="vue--select__option" value="3">Option 3</option>
    </select>
    <chevrondown-stub title="Chevron Down icon" fillcolor="currentColor" size="24" class="vue--select__arrow"></chevrondown-stub>
    <div class="vue--select__selected-value">Option 1</div>
  </div>
  <!---->
</layoutformelement-stub>
`;

exports[`BaseSelect With note displays error variant 1`] = `
<layoutformelement-stub class="vue--select vue--select--default">
  <div class="vue--select__container"><select id="id-drop-mic" name="name-drop-mic" class="vue--select__field vue--select__field--error">
      <option class="vue--select__option" value="1">Option 1</option>
      <option class="vue--select__option" value="2">Option 2</option>
      <option class="vue--select__option" value="3">Option 3</option>
    </select>
    <chevrondown-stub title="Chevron Down icon" fillcolor="currentColor" size="24" class="vue--select__arrow"></chevrondown-stub>
    <!---->
  </div>
  <baseinputnote-stub variant="error" class="vue--select__note">
    Note
  </baseinputnote-stub>
</layoutformelement-stub>
`;

exports[`BaseSelect With note renders as expected 1`] = `
<layoutformelement-stub class="vue--select vue--select--default">
  <div class="vue--select__container"><select id="id-drop-mic" name="name-drop-mic" class="vue--select__field">
      <option class="vue--select__option" value="1">Option 1</option>
      <option class="vue--select__option" value="2">Option 2</option>
      <option class="vue--select__option" value="3">Option 3</option>
    </select>
    <chevrondown-stub title="Chevron Down icon" fillcolor="currentColor" size="24" class="vue--select__arrow"></chevrondown-stub>
    <!---->
  </div>
  <baseinputnote-stub variant="default" class="vue--select__note">
    Note
  </baseinputnote-stub>
</layoutformelement-stub>
`;
